tests/test_misc.lua

```diff
@@ -541,4 +542,382 @@ T['zoom()']['respects `config` argument'] = function()
+end
+
+T['restore_cursor()']['nocenter'] = function()
+  local opts = {
+    ignore_filetype = {},
+    ignore_buftype = {},
+    center = false,
+  }
+
+  child.cmd('edit ' .. test_file_cursor)
+
+  local jump_line = 32
+  cursor_jump_line(jump_line)
+  write_shada_file()
+  child.cmd('bdelete')
+
+  read_shada_file()
+  child.cmd('edit ' .. test_file_cursor)
+  restore_cursor(opts)
+  eq(child.fn.line('.'), jump_line)
+
+  -- For jumps in the middle of the file with enough space, neovim centers automatically.
+  local win_height = child.api.nvim_win_get_height(0)
+  local win_cursor = child.fn.winline()
+  eq(math.floor(win_height / 2 + 1), win_cursor)
+end
+
+T['restore_cursor()']['nocenter eof'] = function()
+  local opts = {
+    ignore_filetype = {},
+    ignore_buftype = {},
+    center = false,
+  }
+
+  child.cmd('edit ' .. test_file_cursor)
+
+  local jump_line = child.fn.line('$')
+  cursor_jump_line(jump_line)
+  write_shada_file()
+  child.cmd('bdelete')
+
+  read_shada_file()
+  child.cmd('edit ' .. test_file_cursor)
+  restore_cursor(opts)
+  eq(child.fn.line('.'), jump_line)
+
+  -- We should be at the end of the window
+  local win_height = child.api.nvim_win_get_height(0)
+  local win_cursor = child.fn.winline()
+  eq(win_height, win_cursor)
+end
+
+T['restore_cursor()']['folds nocenter'] = function()
+  local opts = {
+    ignore_filetype = {},
+    ignore_buftype = {},
+    center = false,
+  }
+
+  child.cmd('edit ' .. test_file_cursor)
+
+  local jump_line = 15
+  cursor_jump_line(jump_line)
+  child.cmd('2,30 fold') -- fold test1 function
+  -- inside fold
+  eq(child.fn.foldclosed('10'), 2)
+  -- outside fold
+  eq(child.fn.foldclosed('40'), -1)
+  write_shada_file()
+  child.cmd('bdelete')
+
+  read_shada_file()
+  child.cmd('edit ' .. test_file_cursor)
+  eq(child.fn.line('.'), 1)
+  restore_cursor(opts)
+  eq(child.fn.line('.'), jump_line)
+  eq(child.fn.foldclosed('.'), -1)
+  eq(child.fn.foldclosed('10'), -1)
+
+  -- For jumps in the middle of the file with enough space, neovim centers automatically.
+  local win_height = child.api.nvim_win_get_height(0)
+  local win_cursor = child.fn.winline()
+  noeq(math.floor(win_height / 2 + 1), win_cursor)
+end
+
+T['restore_cursor()']['folds nocenter eof'] = function()
+  local opts = {
+    ignore_filetype = {},
+    ignore_buftype = {},
+    center = false,
+  }
+
+  child.cmd('edit ' .. test_file_cursor)
+
+  local jump_line = child.fn.line('$')
+  cursor_jump_line(jump_line)
+  child.cmd(string.format('2,%u fold', jump_line))
+  -- inside fold
+  eq(child.fn.foldclosed('30'), 2)
+  write_shada_file()
+  child.cmd('bdelete')
+
+  read_shada_file()
+  child.cmd('edit ' .. test_file_cursor)
+  eq(child.fn.line('.'), 1)
+  restore_cursor(opts)
+  eq(child.fn.line('.'), jump_line)
+  eq(child.fn.foldclosed('30'), -1)
+
+  local win_height = child.api.nvim_win_get_height(0)
+  local win_cursor = child.fn.winline()
+  eq(win_height, win_cursor)
+end
```

Comment:
Are all of these test cases really necessary? On first glance, these cases should be tested:
- Correctly restores cursor when needed.
- Correctly not restores cursor when not needed (each case for every early return in implementation).
- Correctly opens fold (including correctly centering after, not before).
- Correctly centers when needed (both in the middle of the file and end of file).
- Correctly centers when not needed (`opts.center = false`).

Reply:
I've merged some tests.

Reply:
Should be fine now.